File: CSharp/ByRound/2094/2094_C_Brr_Brrr_Patapim.cs
// Problem: CF 2094 C - Brr Brrr Patapim
// https://codeforces.com/contest/2094/problem/C

/*
 * Problem: C. Brr Brrr Patapim
 * Purpose: Given an n×n grid where G[i,j] = p[i+j], reconstruct the permutation p of length 2*n.
 * Algorithm: 
 *   - For each diagonal (i + j = k), we know that all elements on this diagonal are part of the permutation.
 *   - We iterate through diagonals and collect known values from the grid into array b.
 *   - Unknown positions (where G[i,j] = 0) in b[] are filled with remaining unused numbers from [1, 2*n].
 * Time Complexity: O(n^2) per test case due to scanning the grid and processing diagonals.
 * Space Complexity: O(n^2) for storing the grid, plus O(n) for auxiliary arrays (b and c), overall O(n^2).
 */

using System.Linq;
using System;
public class hello {
  static void Main() {
    var T = int.Parse(Console.ReadLine().Trim());
    while (T-- > 0) {
      var n = int.Parse(Console.ReadLine().Trim());
      getAns(n);
    }
  }
  static void getAns(int n) {
    var a = new int[n, n];
    for (int i = 0; i < n; i++) {
      string[] line = Console.ReadLine().Trim().Split(' ');
      for (int j = 0; j < n; j++)
        a[i, j] = int.Parse(line[j]);
    }
    var b = new int[2 * n]; // Stores the reconstructed permutation indexed by (i + j)
    var hs = Enumerable.Range(1, 2 * n).ToHashSet(); // Set of available numbers [1..2*n]
    for (int i = 0; i < n; i++) {
      for (int j = i; j < n; j++) {
        b[i + j + 1] = a[i, j]; // Fill the result array with known values based on diagonal index
        hs.Remove(a[i, j]); // Remove used numbers from set
      }
    }
    var c = hs.ToArray(); // Remaining unused numbers to fill in zeros
    for (int i = 0; i < 2 * n; i++) {
      if (b[i] == 0)
        b[i] = c[0]; // Assign one of the remaining numbers
    }
    Console.WriteLine(string.Join(" ", b));
  }
}


// https://github.com/VaHiX/codeForces/

File: Python/ByRound/2094/2094_C_Brr_Brrr_Patapim.py
# Problem: CF 2094 C - Brr Brrr Patapim
# https://codeforces.com/contest/2094/problem/C

"""
C. Brr Brrr Patapim

Purpose:
This code recovers a permutation of 2*n integers from an n×n grid where each element G[i][j] corresponds to the (i+j)-th element of the permutation.
The approach uses the fact that elements along diagonals (i + j = constant) can be used to deduce values in the permutation.

Algorithms/Techniques:
- Grid traversal with diagonal grouping
- Set operations to find missing elements
- Reconstruction of permutation using known values

Time Complexity: O(n^2) per test case (for reading grid and processing)
Space Complexity: O(n^2) for storing the grid and intermediate data structures

Input Format:
- First line: number of test cases t
- For each test case:
  - First line: n (size of grid)
  - Next n lines: n integers representing the grid G[i][j]

Output Format:
- For each test case: 2*n integers representing the permutation p[1] to p[2n]
"""

t = int(input())  # Read number of test cases
for _ in range(t):
    n = int(input())  # Read size of the grid
    l = list(map(int, input().split()))  # Read first row of the grid
    
    # Process remaining rows of the grid
    for _ in range(n - 1):
        # Append last element of each subsequent row to the list 'l'
        l.append(int(input().split()[-1]))
    
    # Find missing numbers in range [1, 2*n] that are not present in list 'l'
    missing = set(range(1, 2 * n + 1)) - set(l)
    
    # Add missing elements to the front of the list
    l = list(missing) + l
    
    # Print the result as space-separated integers
    print(" ".join(map(str, l)))


# https://github.com/VaHiX/codeForces/